File: packages/virage-core/src/quality/metrics/chunking.ts
/**
 * Component 1 — Chunking metrics (4 metrics)
 *
 * Cohesion:  mean cosine similarity between sentence embeddings within a chunk.
 * Integrity: fraction of AST/structural nodes fully contained within chunk boundaries.
 * Coherence: cosine similarity between adjacent chunks (non-monotonic target 0.4–0.6).
 * Coverage:  fraction of chunks whose token count falls within [min, max] range.
 *
 * Note: Integrity requires parsed AST node boundaries in chunk metadata.
 * Cohesion/Coherence require an embedder for sentence-level vectors.
 * When AST metadata is absent, Integrity is skipped.
 */

import type { MetricResult } from "../interfaces.js";
import { normalizeMonotonicUp01, normalizeCoherence } from "../scoring.js";

export interface ChunkMetricsInput {
  chunks: Array<{
    denseText: string;
    sparseText?: string;
    metadata?: {
      astNodeCount?: number;
      astNodeCountInBounds?: number;
      tokenCount?: number;
    };
  }>;
  embedFn: (text: string) => Promise<number[]>;
  tokenRangeMin?: number;
  tokenRangeMax?: number;
}

function cosineSim(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

function splitSentences(text: string): string[] {
  const sentences = text.match(/[^.!?\n]+[.!?\n]+/g) ?? [];
  return sentences.map((s) => s.trim()).filter((s) => s.length > 10);
}

/** Cohesion: mean pairwise cosine similarity between sentence embeddings within a chunk. */
async function computeCohesion(
  chunks: ChunkMetricsInput["chunks"],
  embedFn: (text: string) => Promise<number[]>,
  sampleSize: number,
): Promise<number> {
  const sample = chunks.slice(0, sampleSize);
  const cohesions: number[] = [];

  for (const chunk of sample) {
    const sentences = splitSentences(chunk.denseText);
    if (sentences.length < 2) continue;

    const embeds = await Promise.all(sentences.map((s) => embedFn(s)));
    let pairSum = 0;
    let pairCount = 0;
    for (let i = 0; i < embeds.length; i++) {
      for (let j = i + 1; j < embeds.length; j++) {
        pairSum += cosineSim(embeds[i], embeds[j]);
        pairCount++;
      }
    }
    if (pairCount > 0) cohesions.push(pairSum / pairCount);
  }

  return cohesions.length === 0
    ? 0
    : cohesions.reduce((s, v) => s + v, 0) / cohesions.length;
}

/** Integrity: fraction of AST nodes fully contained within chunk boundaries (from metadata). */
function computeIntegrity(chunks: ChunkMetricsInput["chunks"]): number | null {
  let totalNodes = 0;
  let boundedNodes = 0;
  for (const chunk of chunks) {
    const total = chunk.metadata?.astNodeCount;
    const bounded = chunk.metadata?.astNodeCountInBounds;
    if (total != null && bounded != null) {
      totalNodes += total;
      boundedNodes += bounded;
    }
  }
  if (totalNodes === 0) return null;
  return boundedNodes / totalNodes;
}

/** Coherence: mean cosine similarity between adjacent chunk embeddings. */
async function computeCoherence(
  chunks: ChunkMetricsInput["chunks"],
  embedFn: (text: string) => Promise<number[]>,
  sampleSize: number,
): Promise<number> {
  if (chunks.length < 2) return 0.5;
  const sample = chunks.slice(0, sampleSize + 1);
  const embeds = await Promise.all(sample.map((c) => embedFn(c.denseText)));
  const sims: number[] = [];
  for (let i = 0; i < embeds.length - 1; i++) {
    sims.push(cosineSim(embeds[i], embeds[i + 1]));
  }
  return sims.reduce((s, v) => s + v, 0) / sims.length;
}

/** Coverage: fraction of chunks whose token count is within [min, max]. */
function computeCoverage(
  chunks: ChunkMetricsInput["chunks"],
  minTokens: number,
  maxTokens: number,
): number {
  if (chunks.length === 0) return 0;
  const inRange = chunks.filter((c) => {
    const tokenCount =
      c.metadata?.tokenCount ?? Math.round(c.denseText.split(/\s+/).length);
    return tokenCount >= minTokens && tokenCount <= maxTokens;
  });
  return inRange.length / chunks.length;
}

export async function computeChunkingMetrics(
  input: ChunkMetricsInput,
  sampleSize = 100,
  weightOverrides: Partial<Record<string, number>> = {},
): Promise<MetricResult[]> {
  const { chunks, embedFn, tokenRangeMin = 50, tokenRangeMax = 512 } = input;
  const sample = chunks.slice(0, sampleSize);

  const [cohesionRaw, coherenceRaw] = await Promise.all([
    computeCohesion(sample, embedFn, sampleSize),
    computeCoherence(sample, embedFn, sampleSize),
  ]);

  const integrityRaw = computeIntegrity(chunks);
  const coverageRaw = computeCoverage(chunks, tokenRangeMin, tokenRangeMax);

  const results: MetricResult[] = [
    {
      name: "Cohesion",
      rawValue: cohesionRaw,
      normalizedValue: normalizeMonotonicUp01((cohesionRaw + 1) / 2),
      weight: weightOverrides["cohesion"] ?? 2.0,
      skipped: false,
    },
    {
      name: "Integrity",
      rawValue: integrityRaw ?? 0,
      normalizedValue:
        integrityRaw != null ? normalizeMonotonicUp01(integrityRaw) : 0,
      weight: weightOverrides["integrity"] ?? 2.0,
      skipped: integrityRaw == null,
      skipReason:
        integrityRaw == null
          ? "AST node boundary metadata not present in chunks"
          : undefined,
    },
    {
      name: "Coherence",
      rawValue: coherenceRaw,
      normalizedValue: normalizeCoherence(coherenceRaw),
      weight: weightOverrides["coherence"] ?? 0.5,
      skipped: false,
    },
    {
      name: "Coverage",
      rawValue: coverageRaw,
      normalizedValue: normalizeMonotonicUp01(coverageRaw),
      weight: weightOverrides["coverage"] ?? 1.0,
      skipped: false,
    },
  ];

  return results;
}

File: crates/virage-engine/src/quality/metrics/chunking.rs
//! Component 1 — Chunking metrics (4 metrics).
//!
//! Cohesion:  mean cosine similarity between sentence embeddings within a chunk.
//! Integrity: fraction of AST/structural nodes fully contained within chunk boundaries.
//! Coherence: cosine similarity between adjacent chunks (non-monotonic target 0.4-0.6).
//! Coverage:  fraction of chunks whose token count falls within [min, max] range.
//!
//! Ported from `dist/quality/metrics/chunking.js` — IR-038.

use std::sync::Mutex;

use crate::embedders::Embedder;
use crate::quality::scoring::{normalize_coherence, normalize_monotonic_up01};
use crate::quality::{MetricResult, QualityChunk};

use super::cosine_sim;

fn split_sentences(text: &str) -> Vec<String> {
    let mut sentences = Vec::new();
    let mut current = String::new();
    for ch in text.chars() {
        current.push(ch);
        if matches!(ch, '.' | '!' | '?' | '\n') {
            let trimmed = current.trim().to_string();
            if trimmed.len() > 10 {
                sentences.push(trimmed);
            }
            current.clear();
        }
    }
    let trimmed = current.trim().to_string();
    if trimmed.len() > 10 {
        sentences.push(trimmed);
    }
    sentences
}

fn embed_texts(
    embedder: &Mutex<dyn Embedder + Send>,
    texts: &[String],
) -> anyhow::Result<Vec<Vec<f32>>> {
    if texts.is_empty() {
        return Ok(Vec::new());
    }
    let text_refs: Vec<&str> = texts.iter().map(String::as_str).collect();
    let mut guard = embedder
        .lock()
        .map_err(|_| anyhow::anyhow!("embedder lock poisoned"))?;
    let dims = guard.dimensions();
    let flat = guard
        .embed_batch(&text_refs)
        .map_err(|e| anyhow::anyhow!("embed error: {e}"))?;
    Ok(flat.chunks(dims).map(|c| c.to_vec()).collect())
}

fn compute_cohesion(
    chunks: &[QualityChunk],
    embedder: &Mutex<dyn Embedder + Send>,
) -> anyhow::Result<f64> {
    let mut cohesions = Vec::new();
    for chunk in chunks {
        let sentences = split_sentences(&chunk.dense_text);
        if sentences.len() < 2 {
            continue;
        }
        let embeds = embed_texts(embedder, &sentences)?;
        let mut pair_sum = 0.0;
        let mut pair_count = 0;
        for i in 0..embeds.len() {
            for j in (i + 1)..embeds.len() {
                pair_sum += cosine_sim(&embeds[i], &embeds[j]);
                pair_count += 1;
            }
        }
        if pair_count > 0 {
            cohesions.push(pair_sum / pair_count as f64);
        }
    }
    if cohesions.is_empty() {
        Ok(0.0)
    } else {
        Ok(cohesions.iter().sum::<f64>() / cohesions.len() as f64)
    }
}

fn compute_integrity(chunks: &[QualityChunk]) -> Option<f64> {
    let mut total_nodes = 0.0;
    let mut bounded_nodes = 0.0;
    for chunk in chunks {
        let total = crate::quality::meta_f64(&chunk.metadata, "astNodeCount");
        let bounded = crate::quality::meta_f64(&chunk.metadata, "astNodeCountInBounds");
        if let (Some(total), Some(bounded)) = (total, bounded) {
            total_nodes += total;
            bounded_nodes += bounded;
        }
    }
    if total_nodes == 0.0 {
        None
    } else {
        Some(bounded_nodes / total_nodes)
    }
}

fn compute_coherence(
    chunks: &[QualityChunk],
    embedder: &Mutex<dyn Embedder + Send>,
) -> anyhow::Result<f64> {
    if chunks.len() < 2 {
        return Ok(0.5);
    }
    let texts: Vec<String> = chunks.iter().map(|c| c.dense_text.clone()).collect();
    let embeds = embed_texts(embedder, &texts)?;
    let mut sims = Vec::new();
    for i in 0..embeds.len().saturating_sub(1) {
        sims.push(cosine_sim(&embeds[i], &embeds[i + 1]));
    }
    if sims.is_empty() {
        Ok(0.0)
    } else {
        Ok(sims.iter().sum::<f64>() / sims.len() as f64)
    }
}

fn estimate_tokens(text: &str) -> f64 {
    text.split_whitespace().count() as f64
}

fn compute_coverage(chunks: &[QualityChunk], min_tokens: f64, max_tokens: f64) -> f64 {
    if chunks.is_empty() {
        return 0.0;
    }
    let in_range = chunks
        .iter()
        .filter(|c| {
            let token_count = crate::quality::meta_f64(&c.metadata, "estimatedTokens")
                .unwrap_or_else(|| estimate_tokens(&c.dense_text));
            token_count >= min_tokens && token_count <= max_tokens
        })
        .count();
    in_range as f64 / chunks.len() as f64
}

pub fn compute_chunking_metrics(
    chunks: &[QualityChunk],
    embedder: &Mutex<dyn Embedder + Send>,
    sample_size: usize,
    token_range_min: f64,
    token_range_max: f64,
) -> anyhow::Result<Vec<MetricResult>> {
    let sample: Vec<QualityChunk> = chunks.iter().take(sample_size).cloned().collect();
    let cohesion_raw = compute_cohesion(&sample, embedder)?;
    let coherence_raw = compute_coherence(&sample, embedder)?;
    let integrity_raw = compute_integrity(chunks);
    let coverage_raw = compute_coverage(chunks, token_range_min, token_range_max);

    let results = vec![
        MetricResult::new(
            "Cohesion",
            cohesion_raw,
            normalize_monotonic_up01((cohesion_raw + 1.0) / 2.0),
            2.0,
        ),
        match integrity_raw {
            Some(v) => MetricResult::new("Integrity", v, normalize_monotonic_up01(v), 2.0),
            None => MetricResult::skipped(
                "Integrity",
                2.0,
                "AST node boundary metadata not present in chunks",
            ),
        },
        MetricResult::new(
            "Coherence",
            coherence_raw,
            normalize_coherence(coherence_raw),
            0.5,
        ),
        MetricResult::new(
            "Coverage",
            coverage_raw,
            normalize_monotonic_up01(coverage_raw),
            1.0,
        ),
    ];
    Ok(results)
}
